Workaround occasional black snapshots on macOS

smfr opened the pull request:
#### c2f8b499772412dbf100797162778234cd634a67
<pre>
Workaround occasional black snapshots on macOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=248718">https://bugs.webkit.org/show_bug.cgi?id=248718</a>
rdar://89840327

Reviewed by Jonathan Bedard.

An OS issue (rdar://89840327) causes test snapshots to be black sometimes,
on Intel machines.

Try to work around this by detecting black snapshots, and re-taking the snapshot
up to three times. We detect black snapshots via their hash, hardcoding
the hashes of some commonly sized black snapshots for performance. The code
logs when black snapshots are detected, but this does not trigger test failure.

Add an all-black ref test in fast/harness to test that all black ref tests
still succeed.

* LayoutTests/fast/harness/all-black-expected.html: Added.
* LayoutTests/fast/harness/all-black.html: Added.
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::compareActualHashToExpectedAndDumpResults):
* Tools/WebKitTestRunner/TestInvocation.h:
* Tools/WebKitTestRunner/cg/TestInvocationCG.cpp:
(WTR::computeSHA1HashStringForContext):
(WTR::operator==):
(WTR::hashForBlackImageOfSize):
(WTR::dumpBitmap):
(WTR::TestInvocation::dumpPixelsAndCompareWithExpected):

Canonical link: <a href="https://commits.webkit.org/257420@main">https://commits.webkit.org/257420@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5b1a1dc66f9b7e689191a7f16615ec0ba5b828be

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98798 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8012 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/31944 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108218 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168473 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102751 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8557 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85380 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91331 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106132 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104481 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/6481 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90038 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33452 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/88303 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21380 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/76375 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/1925 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/22911 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1834 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/45380 "Found 1 new test failure: media/video-inactive-playback.html (failure)") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5108 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6790 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42366 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3230 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard commented on Tools/WebKitTestRunner/cg/TestInvocationCG.cpp:
```diff
@@ -99,15 +100,48 @@ void computeSHA1HashStringForContext(CGContextRef bitmapContext, char hashString
+    if (!context)
+        return { };
+
+    CGContextSetRGBFillColor(context.get(), 0, 0, 0, 1);
```
How often do we hit the non-standard size case?

Might consider a single-variable cache of the most recent non-standard size, which would, I think improve things if the non-standard size is relatively common.

smfr replied:
I figured that a single WebKitTestRunner instance wouldn't benefit much from this, but maybe. I should add logging for the non-standard size case.

smfr committed:
Workaround occasional black snapshots on macOS
https://bugs.webkit.org/show_bug.cgi?id=248718
rdar://89840327

Reviewed by Jonathan Bedard.

An OS issue (rdar://89840327) causes test snapshots to be black sometimes,
on Intel machines.

Try to work around this by detecting black snapshots, and re-taking the snapshot
up to three times. We detect black snapshots via their hash, hardcoding
the hashes of some commonly sized black snapshots for performance. The code
logs when black snapshots are detected, but this does not trigger test failure.

Add an all-black ref test in fast/harness to test that all black ref tests
still succeed.

* LayoutTests/fast/harness/all-black-expected.html: Added.
* LayoutTests/fast/harness/all-black.html: Added.
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::compareActualHashToExpectedAndDumpResults):
* Tools/WebKitTestRunner/TestInvocation.h:
* Tools/WebKitTestRunner/cg/TestInvocationCG.cpp:
(WTR::computeSHA1HashStringForContext):
(WTR::operator==):
(WTR::hashForBlackImageOfSize):
(WTR::dumpBitmap):
(WTR::TestInvocation::dumpPixelsAndCompareWithExpected):

Canonical link: https://commits.webkit.org/257420@main
